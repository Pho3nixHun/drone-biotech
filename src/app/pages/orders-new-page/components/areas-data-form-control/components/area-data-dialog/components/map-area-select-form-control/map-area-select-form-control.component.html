<div class="flex flex-col space-y-4 bg-white">
    <div
        #mapCanvas
        [appElementRef]="mapInitialize"
        class="w-full h-[40vh]"
    ></div>
    <button
        (click)="drawPolygon(null)"
        class="btn"
        [ngClass]="[targetArea() ? 'btn-disabled' : 'btn-primary']"
        [disabled]="targetArea() ? true : false"
    >
        {{ vm().addButtonTextKey | transloco }}
    </button>
    <button
        (click)="deletePolygon()"
        class="btn"
        [ngClass]="[targetArea() ? 'btn-error' : 'btn-disabled']"
        [disabled]="targetArea() ? false : true"
    >
        {{ vm().deleteButtonTextKey | transloco }}
    </button>
</div>
