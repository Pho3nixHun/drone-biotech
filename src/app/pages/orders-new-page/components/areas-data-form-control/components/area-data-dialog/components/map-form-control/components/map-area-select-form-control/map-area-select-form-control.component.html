@let vM = vm();

<div class="flex flex-col space-y-4 bg-white" *transloco="let t">
    <div class="flex flex-col w-1/2">
        <label for="coordinates">{{ t(vM.coordinatesLabelKey) }}</label>
        <textarea
            [id]="coordinatesControl"
            [formControl]="coordinatesControl"
            class="p-2 border rounded-md min-h-32"
            [ngClass]="{
                'border-red-200 bg-red-100':
                    coordinatesControl.invalid && coordinatesControl.dirty,
            }"
        ></textarea>
    </div>

    <div class="join">
        <app-button
            [vm]="vM.addButtonVM"
            (clicked)="drawPolygon(null)"
            [state]="polygonPathsSignal() ? 'disabled' : 'default'"
        />

        <app-button
            [vm]="vM.editButtonVM"
            (clicked)="editPolygon()"
            [state]="canBeEdited() ? 'default' : 'disabled'"
        />

        <app-button
            [vm]="vM.deleteButtonVM"
            (clicked)="deletePolygon()"
            [state]="polygonPathsSignal() ? 'default' : 'disabled'"
        />
    </div>
</div>
