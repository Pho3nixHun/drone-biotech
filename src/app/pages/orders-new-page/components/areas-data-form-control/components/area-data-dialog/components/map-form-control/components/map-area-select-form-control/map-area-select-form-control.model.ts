import { InjectionToken } from '@angular/core';

export interface MapAreaSelectFormControlVM {
    areaValueKey: string;
    addButtonTextKey: string;
    editButtonTextKey: string;
    deleteButtonTextKey: string;
    coordinatesLabelKey: string;
}

export const POLYGON_OPTIONS = new InjectionToken<google.maps.PolygonOptions>(
    'Injection Token for Polygon options'
);

export const provideMockPolygonOptions = () => ({
    provide: POLYGON_OPTIONS,
    useValue: {},
});
export const INFO_WINDOW_OPTIONS =
    new InjectionToken<google.maps.InfoWindowOptions>(
        'Injection Token for Info Window options'
    );

export const provideMockInfoWindowOptions = () => ({
    provide: INFO_WINDOW_OPTIONS,
    useValue: {},
});
