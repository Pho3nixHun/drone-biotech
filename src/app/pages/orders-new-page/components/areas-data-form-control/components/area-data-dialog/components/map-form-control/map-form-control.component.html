<div class="flex flex-col space-y-2">
    <app-map-search-input-form-control
        [defaultCenter]="vm().defaultCenter"
        [vm]="vm().mapSearchInputFormControlVM"
        (changeBounds)="setMapBounds($event)"
    />

    <div appElementRef [signal]="divSignal" style="height: 600px"></div>

    <app-map-area-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapAreaSelectFormControlVM"
        [formControl]="mapForm.controls.targetArea"
    />
    <app-map-point-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapPointSelectFormControlVM"
        [formControl]="mapForm.controls.entryPoint"
    />
</div>
