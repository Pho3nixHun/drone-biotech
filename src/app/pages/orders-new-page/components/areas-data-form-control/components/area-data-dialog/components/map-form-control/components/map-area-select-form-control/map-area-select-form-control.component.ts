import {
    Component,
    computed,
    effect,
    forwardRef,
    inject,
    input,
} from '@angular/core';
import { MapAreaSelectFormControlVM } from './map-area-select-form-control.model';
import {
    Coordinates,
    SQUARE_METRES_TO_HECTARE,
} from '@stores/location/location.model';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
    Validators,
} from '@angular/forms';
import { MapAreaSelectFormControlService } from './map-area-select-form-control.service';
import { TranslocoModule, TranslocoService } from '@jsverse/transloco';
import { toSignal } from '@angular/core/rxjs-interop';
import { NgClass } from '@angular/common';
import { coordinatesStringValidator } from '@utils/array-length.validator';
import { getAreaOfPolygon, getCenter } from 'geolib';
import {
    mapCoordinatesToString,
    mapStringToCoordinates,
} from '../../map-form-control.model';
import { map } from 'rxjs';
import { POLYGON_OPTIONS } from '@providers/google-maps-provider';

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {};

@Component({
    selector: 'app-map-area-select-form-control',
    imports: [TranslocoModule, NgClass, ReactiveFormsModule],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapAreaSelectFormControlComponent),
            multi: true,
        },
        {
            provide: MapAreaSelectFormControlService,
            deps: [POLYGON_OPTIONS],
        },
    ],
    templateUrl: './map-area-select-form-control.component.html',
})
export class MapAreaSelectFormControlComponent implements ControlValueAccessor {
    private readonly translocoService = inject(TranslocoService);
    public readonly mapRef = input.required<google.maps.Map | null>();
    public readonly vm = input.required<MapAreaSelectFormControlVM>();
    private readonly fb = inject(FormBuilder);
    private readonly areaSelectService = inject(
        MapAreaSelectFormControlService
    );

    private readonly initializeMapEffect = effect(
        () => {
            const map = this.mapRef();
            if (!map) return;
            this.areaSelectService.initializeMap(map);
            this.initializeMapEffect.destroy();
        },
        { manualCleanup: true }
    );

    protected readonly polygonPathsSignal = toSignal(
        this.areaSelectService.polygonPaths$,
        { initialValue: null }
    );

    private readonly polygonCoordinatesEffect = effect(() =>
        this.onChange(this.polygonPathsSignal())
    );

    private readonly coordinatesToString = computed(() => {
        const paths = this.polygonPathsSignal();
        return paths ? mapCoordinatesToString(paths) : '';
    });

    private readonly setCoordinatesControlValueEffect = effect(() =>
        this.coordinatesControl.setValue(this.coordinatesToString())
    );

    protected readonly coordinatesControl = this.fb.control<string>('', [
        Validators.required,
        coordinatesStringValidator(3),
    ]);

    private readonly coordinatesFormControlValueSignal = toSignal(
        this.coordinatesControl.valueChanges.pipe(map((data) => data ?? '')),
        { initialValue: '' }
    );

    protected readonly canBeEdited = computed(() => {
        const coords = this.coordinatesToString();
        const control = this.coordinatesFormControlValueSignal();

        if (this.coordinatesControl.invalid) return false;
        return coords !== control;
    });

    private readonly setInfoWindowEffect = effect(() => {
        const center = this.polygonCenterSignal();
        const size = this.areSizeInHaSignal();
        if (!center || !size) return;

        this.areaSelectService.setInfoWindow(
            this.translocoService.translate(this.vm().areaValueKey, {
                value: size,
            }),
            center
        );
    });

    private readonly polygonCenterSignal = computed(() => {
        const polygon = this.polygonPathsSignal();
        if (!polygon) return null;
        const center = getCenter(polygon);

        if (!center) return null;

        return { lat: center.latitude, lng: center.longitude };
    });

    private readonly areSizeInHaSignal = computed(() => {
        const polygon = this.polygonPathsSignal();
        return polygon
            ? getAreaOfPolygon(polygon) / SQUARE_METRES_TO_HECTARE
            : null;
    });

    protected drawPolygon(coordinates: Coordinates[] | null) {
        this.areaSelectService.drawPolygon(coordinates);
    }

    protected deletePolygon() {
        this.areaSelectService.deletePolygon();
    }
    protected editPolygon() {
        if (this.coordinatesControl.valid)
            this.areaSelectService.editPolygon(
                mapStringToCoordinates(this.coordinatesControl.value ?? '')
            );
    }

    private onChange: (value: Coordinates[] | null) => void = noop;

    private onTouched: () => void = noop;

    public writeValue(value: Coordinates[] | null) {
        if (value) return this.areaSelectService.drawPolygon(value);
        this.areaSelectService.deletePolygon();
    }

    public registerOnChange(fn: (value: Coordinates[] | null) => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
