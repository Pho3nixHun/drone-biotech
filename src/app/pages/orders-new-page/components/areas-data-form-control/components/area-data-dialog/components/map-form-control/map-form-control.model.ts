import { MapSearchInputFormControlVM } from 'src/app/pages/orders-new-page/components/areas-data-form-control/components/area-data-dialog/components/map-form-control/components/map-search-input-form-control/map-search-input-form-control.model';
import { MapAreaSelectFormControlVM } from './components/map-area-select-form-control/map-area-select-form-control.model';
import { MapPointSelectFormControlVM } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { Coordinates } from '../../area-data-dialog.model';
import { InjectionToken } from '@angular/core';

export interface MapFormControlVM {
    defaultCenter: Coordinates | null;
    mapAreaSelectFormControlVM: MapAreaSelectFormControlVM;
    mapPointSelectFormControlVM: MapPointSelectFormControlVM;
    mapSearchInputFormControlVM: MapSearchInputFormControlVM;
}

export const MAP_OPTIONS = new InjectionToken<MapOptions>(
    'Injection Token for Map options'
);

export const provideMockMapOptions = () => ({
    provide: MAP_OPTIONS,
    useValue: {},
});
export interface MapOptions {
    mapTypeId: 'roadmap';
    disableDoubleClickZoom: boolean;
    clickableIcons: boolean;
    mapId: 'DEMO_MAP_ID';
    zoom: number;
    streetViewControl: boolean;
}

export const mapStringToCoordinates = (input: string): Coordinates[] => {
    const lines = input
        .split('\n')
        .map((line) => line.trim())
        .filter((line) => line !== ''); // ignore empty lines

    const coordinates: Coordinates[] = lines.map((line) => {
        const [lat, lng] = line.split(' ').map(Number);
        return { lat, lng };
    });

    return coordinates;
};

export const mapCoordinatesToString = (coordinates: Coordinates[]): string => {
    return coordinates.map((coord) => `${coord.lat} ${coord.lng}`).join('\n');
};
