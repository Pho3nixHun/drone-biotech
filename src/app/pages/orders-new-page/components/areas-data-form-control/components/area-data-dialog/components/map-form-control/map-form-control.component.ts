/* eslint-disable @typescript-eslint/no-empty-function */
import {
    Component,
    computed,
    effect,
    ElementRef,
    forwardRef,
    inject,
    input,
    Signal,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
    Validators,
} from '@angular/forms';
import { MapFormControlVM } from './map-form-control.model';
import { Coordinates } from '@stores/location/location.model';
import { MapAreaSelectFormControlComponent } from './components/map-area-select-form-control/map-area-select-form-control.component';
import { MapPointSelectFormControlComponent } from './components/map-point-select-form-control/map-point-select-form-control.component';
import { MapSearchInputFormControlComponent } from './components/map-search-input-form-control/map-search-input-form-control.component';
import { distinctUntilChanged, map } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';
import { getCenter } from 'geolib';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import {
    MAP_OPTIONS,
    HEAD_OFFICE_LOCATION,
} from '@providers/google-maps-provider';

@Component({
    selector: 'app-map-form-control',
    imports: [
        MapAreaSelectFormControlComponent,
        MapPointSelectFormControlComponent,
        ElementRefDirective,
        ReactiveFormsModule,
        MapSearchInputFormControlComponent,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    schemas: [],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    private readonly mapOptions = inject(MAP_OPTIONS);
    protected readonly headOfficeLocation = inject(HEAD_OFFICE_LOCATION);
    public readonly vm = input.required<MapFormControlVM>();
    protected readonly divSignal = signal<ElementRef<HTMLElement> | null>(null);
    private readonly targetArea = signal<Coordinates[] | null>(null);

    protected readonly mapSignal = computed(() => {
        const canvas = this.divSignal();
        if (!canvas) return null;
        return new google.maps.Map(canvas.nativeElement);
    });

    private readonly targetAreaCenter = computed<Coordinates | null>(() => {
        const area = this.targetArea();
        if (!area) return null;
        const center = getCenter(area);
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    protected readonly mapCenter = computed<Coordinates>(
        () =>
            this.targetAreaCenter() ??
            this.vm().defaultCenter ??
            this.headOfficeLocation
    );

    protected readonly setMapCenterEffect = effect(
        () => {
            const map = this.mapSignal();
            if (!map) return;
            map.setCenter(this.mapCenter());
            this.setMapCenterEffect.destroy();
        },
        { manualCleanup: true }
    );

    protected mapForm = this.fb.group({
        targetArea: this.fb.control<Coordinates[] | null>(
            null,
            Validators.required
        ),
        entryPoint: this.fb.control<Coordinates | null>(
            null,
            Validators.required
        ),
    });

    private readonly setMapOptionsEffect = effect(() => {
        const map = this.mapSignal();
        if (!map) return;

        map.setOptions({
            ...this.mapOptions,
            mapTypeControlOptions: {
                position: google.maps.ControlPosition.TOP_RIGHT,
            },
        });
    });

    protected setMapBounds(bounds: google.maps.LatLngBounds | null) {
        const map = this.mapSignal();
        if (!bounds || !map) return;
        map.fitBounds(bounds);
    }

    private readonly mapFormValue: Signal<{
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }> = toSignal(
        this.mapForm.valueChanges.pipe(
            distinctUntilChanged(),
            map((data) => {
                const { entryPoint, targetArea } = data;
                return {
                    entryPoint: entryPoint ?? null,
                    targetArea: targetArea ?? null,
                };
            })
        ),
        { initialValue: { targetArea: null, entryPoint: null } }
    );

    private readonly mapFormValueEffect = effect(() =>
        this.onChange(this.mapFormValue())
    );

    private onChange: (value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) => void = () => {};

    private onTouched: () => void = () => {};

    public writeValue(value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) {
        this.mapForm.setValue(value);
        this.targetArea.set(value.targetArea);
    }

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
