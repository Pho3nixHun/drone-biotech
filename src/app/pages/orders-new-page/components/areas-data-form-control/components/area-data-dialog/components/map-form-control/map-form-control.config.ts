import { EnvironmentProviders, Provider } from '@angular/core';
import { HEAD_OFFICE_LOCATION } from '@services/distance/distance.model';
import {
    GOOGLE_MAPS_LOADER_CONFIG,
    provideGoogleMapsLibraries,
} from 'src/app/shared/providers/google-maps-provider';
import { environment } from 'src/environments/environment';
import {
    POLYGON_OPTIONS,
    INFO_WINDOW_OPTIONS,
} from './components/map-area-select-form-control/map-area-select-form-control.model';
import { ENTRY_POINT_MARKER_OPTIONS } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { MAP_OPTIONS } from './map-form-control.model';

export const mapFormControlConfig: (Provider | EnvironmentProviders)[] = [
    provideGoogleMapsLibraries(),
    {
        provide: POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: true,
            fillOpacity: 0.5,
            strokeColor: 'blue',
            fillColor: 'blue',
        },
    },
    {
        provide: MAP_OPTIONS,
        useValue: {
            mapTypeId: 'roadmap',
            disableDoubleClickZoom: true,
            clickableIcons: false,
            mapId: 'DEMO_MAP_ID',
            zoom: 14,
            streetViewControl: false,
        },
    },
    {
        provide: ENTRY_POINT_MARKER_OPTIONS,
        useValue: {
            gmpDraggable: true,
        },
    },
    {
        provide: GOOGLE_MAPS_LOADER_CONFIG,
        useValue: {
            url: 'https://maps.googleapis.com/maps/api/js?',
            key: environment.googleMapsConfig.apiKey,
            version: '3.59',
            region: 'HU',
        },
    },
    {
        provide: HEAD_OFFICE_LOCATION,
        useValue: { lat: 47.312498121576795, lng: 21.309304570654604 },
    },
    {
        provide: INFO_WINDOW_OPTIONS,
        useValue: {
            disableAutoPan: true,
            headerDisabled: true,
        },
    },
];
