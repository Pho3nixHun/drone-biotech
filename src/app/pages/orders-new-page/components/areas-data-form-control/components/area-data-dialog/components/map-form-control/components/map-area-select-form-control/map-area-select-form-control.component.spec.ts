import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MapAreaSelectFormControlComponent } from './map-area-select-form-control.component';
import { provideMockHeadOfficeLocation } from '@services/distance/distance.model';
import {
    MapAreaSelectFormControlVM,
    provideMockInfoWindowOptions,
    provideMockPolygonOptions,
} from './map-area-select-form-control.model';
import { Coordinates } from '@stores/location/location.model';
import { getTranslocoModule } from 'transloco-testing.module';
import {
    Component,
    computed,
    DebugElement,
    ElementRef,
    inject,
    input,
    signal,
} from '@angular/core';
import { MapAreaSelectFormControlService } from './map-area-select-form-control.service';
import {
    MapAreaSelectFormControlMockService,
    updatePolygonSignal,
} from './map-area-select-form-control.service.mock';
import { By } from '@angular/platform-browser';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import { provideMockMapOptions } from '../../map-form-control.model';
import { FormBuilder, ReactiveFormsModule } from '@angular/forms';
import { MatIcon } from '@interfaces/mat-icon.enum';

const enMock = {
    areaValue: 'area',
    coordinatesLabel: 'coordinates',
};

const vm: MapAreaSelectFormControlVM = {
    addButtonVM: { type: 'withIcon', icon: MatIcon.ADD, variant: 'fill' },
    deleteButtonVM: { type: 'withIcon', icon: MatIcon.ADD, variant: 'fill' },
    editButtonVM: { type: 'withIcon', icon: MatIcon.ADD, variant: 'fill' },
    areaValueKey: enMock.areaValue,
    coordinatesLabelKey: enMock.coordinatesLabel,
};
@Component({
    imports: [
        MapAreaSelectFormControlComponent,
        ReactiveFormsModule,
        ElementRefDirective,
    ],
    template: ` <div
            appElementRef
            [signal]="divSignal"
            style="height: 600px"
        ></div>

        <app-map-area-select-form-control
            [mapRef]="mapSignal()"
            [vm]="vm()"
            [formControl]="control"
        />`,
})
class TestHostComponent {
    private readonly fb = inject(FormBuilder);
    public vm = input.required<MapAreaSelectFormControlVM>();
    public control = this.fb.control<Coordinates[] | null>([]);
    protected readonly divSignal = signal<ElementRef<HTMLElement> | null>(null);
    protected readonly mapSignal = computed(() => {
        const canvas = this.divSignal();
        if (!canvas) return null;
        return new google.maps.Map(canvas.nativeElement);
    });
}

describe('MapAreaSelectFormControlComponent', () => {
    let fixture: ComponentFixture<TestHostComponent>;
    let component: TestHostComponent;
    let mapAreaSelectFormControlComponent: MapAreaSelectFormControlComponent;
    let innerDebugElement: DebugElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                TestHostComponent,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [
                provideMockHeadOfficeLocation(),
                provideMockMapOptions(),
                provideMockInfoWindowOptions(),
                provideMockPolygonOptions(),
            ],
        }).compileComponents();
        TestBed.overrideProvider(MapAreaSelectFormControlService, {
            useFactory: () => new MapAreaSelectFormControlMockService(),
        });
        fixture = TestBed.createComponent(TestHostComponent);

        component = fixture.componentInstance;
        innerDebugElement = fixture.debugElement.query(
            By.directive(MapAreaSelectFormControlComponent)
        );
        mapAreaSelectFormControlComponent = innerDebugElement.componentInstance;
    });

    // Interaction test
    it('should call drawPolygon(null) when the addButton is clicked if there is no target area drawn', () => {
        //Arrange
        fixture.componentRef.setInput('vm', vm);
        updatePolygonSignal(null);

        //Act
        fixture.detectChanges();
        const serviceSpy = jest.spyOn(
            mapAreaSelectFormControlComponent['areaSelectService'],
            'drawPolygon'
        );
        fixture.debugElement
            .query(By.css('button.btn-primary'))
            .triggerEventHandler('click', null);

        //Assert
        expect(serviceSpy).toHaveBeenCalledWith(null);
    });

    it('should initialize the map with the mapCanvas', () => {
        //Arrange
        fixture.componentRef.setInput('vm', vm);
        const serviceSpy = jest.spyOn(
            mapAreaSelectFormControlComponent['areaSelectService'],
            'initializeMap'
        );

        //Act
        fixture.detectChanges();

        // Assert
        expect(serviceSpy).toHaveBeenCalledWith(component['mapSignal']());
    });
});
