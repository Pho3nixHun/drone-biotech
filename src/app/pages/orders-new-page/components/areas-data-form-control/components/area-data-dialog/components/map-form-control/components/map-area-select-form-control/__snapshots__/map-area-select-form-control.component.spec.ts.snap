// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapAreaSelectFormControlComponent should render the template when the VM is provided 1`] = `
<app-map-area-select-form-control
  class="ng-untouched ng-dirty ng-valid"
  ng-reflect-form="[object Object]"
  ng-reflect-map-ref="[object Object]"
  ng-reflect-vm="[object Object]"
>
  <div
    class="flex flex-col space-y-4 bg-white"
  >
    <div
      class="flex flex-col w-1/2"
    >
      <label
        for="coordinates"
      >
        coordinates
      </label>
      <textarea
        class="p-2 border rounded-md min-h-32 ng-untouched ng-pristine ng-invalid"
        id="[object Object]"
        ng-reflect-form="[object Object]"
        ng-reflect-ng-class="[object Object]"
      />
    </div>
    <button
      class="btn btn-primary"
      ng-reflect-ng-class="btn-primary"
    >
       addButton 
    </button>
    <button
      class="btn btn-disabled"
      disabled=""
      ng-reflect-ng-class="btn-disabled"
    >
       edit 
    </button>
    <button
      class="btn btn-disabled"
      disabled=""
      ng-reflect-ng-class="btn-disabled"
    >
       deleteButton 
    </button>
  </div>
  <!--container-->
</app-map-area-select-form-control>
`;
